File: _courses/cmpt225/lecture09/python/queue.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None


class QueueLinkedList:
    """
    linked list implementation of queue
    - similar to doubly linked list
    """

    def __init__(self):
        self.front = None
        self.back = None

    def push(self, data):
        """
        push element to back of queue, O(1)
        """
        if not self.back:
            self.back = Node(data)
            self.front = self.back
        else:
            node = Node(data)
            self.back.next = node
            node.prev = self.back
            self.back = node

    def pop(self):
        """
        pop element from front of queue, O(1)
        """
        if not self.front:
            raise Exception("Cannot pop from empty queue")
        data = self.front.data
        self.front = self.front.next
        if not self.front:
            self.back = None
        else:
            self.front.prev = None
        return data

    def __str__(self):
        """
        string representation of queue
        """
        s = ""
        node = self.front
        while node is not None:
            s += str(node.data) + " "
            node = node.next
        return s


class QueueArrayList:
    """
    array list implementation of queue
    """

    def __init__(self):
        self.arr = []

    def push(self, data):
        self.arr.append(data)

    def pop(self):
        return self.arr.pop(0)

    def __str__(self):
        return str(self.arr)

    
if __name__ == "__main__":

    queue = QueueLinkedList()
    queue.push(5)
    queue.push(3)
    queue.push(7)
    print(str(queue)) # 5, 3, 7
    queue.pop()
    print(str(queue)) # 3, 7
    queue.pop()
    queue.pop()

    queue = QueueArrayList()
    queue.push(5)
    queue.push(3)
    queue.push(7)
    print(str(queue)) # 5, 3, 7
    queue.pop()
    print(str(queue)) # 3, 7
    queue.pop()
    queue.pop()

File: _courses/cmpt225/lecture09/cpp/src/queue.cpp
#include "queue.h"
#include "node.h"

#include <exception>
#include <string>

template <class Type>
QueueLinkedList<Type>::QueueLinkedList() : front(nullptr), back(nullptr)
{
}

template <class Type>
QueueLinkedList<Type>::~QueueLinkedList()
{
    Node<Type> *node = front;
    while (node)
    {
        Node<Type> *next = node->next;
        delete node;
        node = next;
    }
}

template <class Type>
void QueueLinkedList<Type>::push(Type data)
{
    if (!back)
    {
        back = new Node<Type>(data);
        front = back;
    }
    else
    {
        Node<Type> *node = new Node<Type>(data);
        back->next = node;
        node->prev = back;
        back = node;
    }
}

template <class Type>
Type QueueLinkedList<Type>::pop()
{
    if (!front)
    {
        // "Cannot pop from empty queue"
        throw std::exception();
    }

    Type data = front->data;
    Node<Type> *next = front->next;
    delete front;
    front = next;

    if (!front)
    {
        back = nullptr;
    }
    else
    {
        front->prev = nullptr;
    }

    return data;
}

template <>
std::string QueueLinkedList<int>::to_string() const
{
    std::string s;

    Node<int> *node = front;

    while (node)
    {
        s += std::to_string(node->data) + " ";
        node = node->next;
    }

    return s;
}

template <>
std::string QueueLinkedList<std::string>::to_string() const
{
    std::string s;

    Node<std::string> *node = front;

    while (node)
    {
        s += node->data + " ";
        node = node->next;
    }

    return s;
}

template <class Type>
void QueueArrayList<Type>::push(Type data)
{
    arr.push_back(data);
}

template <class Type>
Type QueueArrayList<Type>::pop()
{
    if (arr.empty())
    {
        // "Cannot pop from empty queue"
        throw std::exception();
    }

    Type data = arr[0];

    arr.erase(arr.begin());

    return data;
}

template <>
std::string QueueArrayList<int>::to_string() const
{
    std::string s;

    for (auto it : arr)
    {
        s += std::to_string(it) + " ";
    }

    return s;
}

template <>
std::string QueueArrayList<std::string>::to_string() const
{
    std::string s;

    for (auto it : arr)
    {
        s += it + " ";
    }

    return s;
}

template class QueueLinkedList<int>;
template class QueueLinkedList<std::string>;

template class QueueArrayList<int>;
template class QueueArrayList<std::string>;